src/zcash/address/unified.cpp

```diff
@@ -75,62 +72,91 @@ ZcashdUnifiedFullViewingKey ZcashdUnifiedFullViewingKey::FromUnifiedFullViewingK
+    if (receiverTypes.count(ReceiverType::P2PKH) > 0) {
+        if (transparentKey.has_value()) {
+            const auto& tkey = transparentKey.value();
 
-        CPubKey externalKey;
-        ChainCode externalChainCode;
-        if (!tkey.GetPubKey().Derive(externalKey, externalChainCode, 0, tkey.GetChainCode())) {
-            return std::nullopt;
-        }
+            auto childIndex = j.ToTransparentChildIndex();
+            if (!childIndex.has_value()) return UnifiedAddressGenerationError::InvalidTransparentChildIndex;
+
+            auto externalPubkey = tkey.DeriveExternal(childIndex.value());
+            if (!externalPubkey.has_value()) return UnifiedAddressGenerationError::NoAddressForDiversifier;
 
-        CPubKey childKey;
-        ChainCode childChainCode;
-        if (externalKey.Derive(childKey, childChainCode, childIndex.value(), externalChainCode)) {
-            ua.AddReceiver(childKey.GetID());
+            ua.AddReceiver(externalPubkey.value().GetID());
         } else {
-            return std::nullopt;
+            return UnifiedAddressGenerationError::ReceiverTypeNotAvailable;
         }
     }
 
-    return ua;
+    return std::make_pair(ua, j);
 }
 
-std::optional<std::pair<UnifiedAddress, diversifier_index_t>> ZcashdUnifiedFullViewingKey::FindAddress(
+UnifiedAddressGenerationResult ZcashdUnifiedFullViewingKey::FindAddress(
         const diversifier_index_t& j,
         const std::set<ReceiverType>& receiverTypes) const {
     diversifier_index_t j0(j);
     bool hasTransparent = HasTransparent(receiverTypes);
     auto addr = Address(j0, receiverTypes);
-    while (!addr.has_value()) {
-        if (!j0.increment() || (hasTransparent && !j0.ToTransparentChildIndex().has_value()))
-            return std::nullopt;
+    while (addr == UnifiedAddressGenerationResult(UnifiedAddressGenerationError::NoAddressForDiversifier)) {
+        if (!j0.increment())
+            return UnifiedAddressGenerationError::DiversifierSpaceExhausted;
+
+        if (hasTransparent && !j0.ToTransparentChildIndex().has_value())
+            return UnifiedAddressGenerationError::InvalidTransparentChildIndex;
+
         addr = Address(j0, receiverTypes);
     }
-    return std::make_pair(addr.value(), j0);
+    return addr;
 }
 
-std::optional<std::pair<UnifiedAddress, diversifier_index_t>> ZcashdUnifiedFullViewingKey::FindAddress(
+UnifiedAddressGenerationResult ZcashdUnifiedFullViewingKey::FindAddress(
         const diversifier_index_t& j) const {
     return FindAddress(j, {ReceiverType::P2PKH, ReceiverType::Sapling});
 }
+
+std::optional<RecipientAddress> ZcashdUnifiedFullViewingKey::GetChangeAddress(const ChangeRequest& req) const {
+    std::optional<RecipientAddress> addr;
+    std::visit(match {
+        [&](const TransparentChangeRequest& req) {
+            if (transparentKey.has_value()) {
+                auto changeAddr = transparentKey.value().GetChangeAddress(req.GetIndex());
```

Comment:
Hmm, so we select from a sequence of transparent change addresses, but only a single shielded change address? That could lead to inconsistencies in UX (compared to external addresses), but I _think_ it's fine as long as change addresses are never exposed directly in UIs (instead _always_ showing the default external address), which is the intention.

Reply:
That would be reasonable if we had a way to import a mnemonic (which we never want to do; what we'd actually want is to start a new `wallet.dat` from that mnemonic). But as `z_importwallet` just imports the keys directly, if we _didn't_ include the change addresses then the user would lose access to funds. Note that the importer _does_ attempt to import the key's `hdKeypath` if it and the `seedFpStr` are available, so the change address keypaths would still be identified. Also, for transparent spending keys, there is a `change=1` flag indicating if they were change addresses (for the legacy randomly-generated ones), so there is precedent for including them. Maybe what we _wouldn't want to do is include the `# zaddr=%s` tail comment for interenal addresses.